### ISSUE TITLE
How to update the parameters βn and vn?

### ISSUE THREAD

**Issue by ChunyanZhao** [layog/Accurate-Binary-Convolution-Network#4]
Hi, @layog 

I am glad to find your code, it is a great help for my understanding of the paper. But I am confused that how do you update the β and ν in activation layer? I notice that you have set the initial value, 1 for β and 0 for ν, is it the exact implementation of the paper which is described as statistical features?

**Comment by layog** [comment]
Yes, the β (called _betas_ in the code) and ν (called _shift_parameters_ in the code) are trainable. Both of these are defined as variables and hence are trained when training the complete network. These along with BatchNorm parameters are trained when _train_step_ is computed. The initial values are chosen to be 1 for _betas_ and 0 for _shift_parameters_ (just like it is done in the BatchNorm layer) in order to retain the actual information and these are subsequently trained by the network.

**Comment by ChunyanZhao** [comment]
`        
def ABCLayer(input_tensor, name=None):
            
         with tf.name_scope(name, "ABCLayer"):

                # Reshaping betas to match the input tensor
                reshaped_betas = tf.reshape(betas,
                                            shape=[no_ApproxConvLayers] + [1] * len(input_tensor.get_shape()),
                                            name="reshaped_betas")

                # Calculating ApproxConv for each shifted input
                ApproxConv_layers = []
                for index in range(no_ApproxConvLayers):
                    # Shifting and binarizing input
                    shifted_input = tf.clip_by_value(input_tensor + shift_parameters[index], 0., 1.,
                                                     name="shifted_input_" + str(index))
                    binarized_activation = tf.sign(shifted_input - 0.5)
                    
                    # Passing through the ApproxConv layer
                    ApproxConv_layers.append(ApproxConvLayer(binarized_activation))
                ApproxConv_output = tf.convert_to_tensor(ApproxConv_layers, dtype=tf.float32,
                                                         name="ApproxConv_output")
                
                # Taking the weighted sum using the betas
                ABC_output = tf.reduce_sum(tf.multiply(ApproxConv_output, reshaped_betas), axis=0)
                return ABC_output`
As is shown above,  shift_parameters and reshaped_betas are computed in the same way with the same data,  I wonder if all shift_parameters will be trained to the same value. Or I miss something?

**Comment by layog** [comment]
You are probably right. This might be a fatal bug in the code, that I have overlooked. Nonetheless, I would have to test this whether all _shift_parameters_ and _betas_ are learning the same values. If you have already tested it, do share your results.

If it actually is a bug, an easy fix would be to initialize _shift_parameters_ and _betas_ from a gaussian (Glorot initialization should be better) rather than fixed values.

P.S. - Thanks for pointing this out

**Comment by ChunyanZhao** [comment]
I'm sorry that I haven't tested it, but I find that the paper gives some advise to initializing _shift_parameters_ and _betas_ in supplementary material. Thank you again for your generous share.
